File: templates/problems/07_trees/09_right_side_view/python/solution.py
from helpers import TreeNode, Optional


def solve_right_side_view(root: Optional[TreeNode]) -> list[int]:
    """Return the values of the nodes visible from the right side.

    Time: O(n), Space: O(n).
    """
    if root is None:
        return []
    result: list[int] = []
    queue: list[TreeNode] = [root]
    while queue:
        size = len(queue)
        for i in range(size):
            node = queue.pop(0)
            if i == size - 1:
                result.append(node.val)
            if node.left is not None:
                queue.append(node.left)
            if node.right is not None:
                queue.append(node.right)
    return result

File: templates/problems/07_trees/09_right_side_view/go/solution.go
package right_side_view

// solveRightSideView returns the values of the nodes visible from the right side.
// Time: O(n), Space: O(n).
func solveRightSideView(root *TreeNode) []int {
	if root == nil {
		return []int{}
	}
	var result []int
	queue := []*TreeNode{root}
	for len(queue) > 0 {
		size := len(queue)
		for i := 0; i < size; i++ {
			node := queue[i]
			if i == size-1 {
				result = append(result, node.Val)
			}
			if node.Left != nil {
				queue = append(queue, node.Left)
			}
			if node.Right != nil {
				queue = append(queue, node.Right)
			}
		}
		queue = queue[size:]
	}
	return result
}
